### PR TITLE
fix: uneven load among clickhouse shards caused by retry error mechanism

### PR DESCRIPTION
**Pull request by nir3c**
## Description

fix uneven load among clickhouse shards caused by the retry error mechanism (related to issues #325 and #322 )

The issue happens when ChProxy sends the HTTP call to one of the shards with a retry mechanism.

When the response from the shard is 502 the [scope's host swaps](https://github.com/ContentSquare/chproxy/blob/6fb82516c6b2066a3a4cd271f22ae741e75b551d/proxy.go#L251) with another one to send by it, which causes not to decreasing the initial host's counter and the new host's counter to decrease instead (when we finally [decrementing the scope](https://github.com/ContentSquare/chproxy/blob/6fb82516c6b2066a3a4cd271f22ae741e75b551d/proxy.go#L107)), which could lead to a situation when the new host's counter is 0 to be set to MaxUint32 value (as `0 - 1` equals to the maximum value of uint32).
once this situation happens the new host will not be able to receive any new requests (as its counter value will be always greater than all other hosts' counter, and the [balancing function logic for getting the new host](https://github.com/ContentSquare/chproxy/blob/6fb82516c6b2066a3a4cd271f22ae741e75b551d/scope.go#L1011) is [based on the host's counter + penalty](https://github.com/ContentSquare/chproxy/blob/6fb82516c6b2066a3a4cd271f22ae741e75b551d/scope.go#L823)) and the host's replica will never be used (same as for the host -> it will always be higher value than all other replicas as we can see in the [getReplica function](https://github.com/ContentSquare/chproxy/blob/6fb82516c6b2066a3a4cd271f22ae741e75b551d/scope.go#L900))

 

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Checklist

- [ ] Linter passes correctly
- [ ] Add tests which fail without the change (if possible)
- [x] All tests passing
- [ ] Extended the README / documentation, if necessary

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Comment by sigua-cs** [comment]
@nir3c thanks for the PR. Please check failed tests

File: proxy.go
Comment (nir3c) [inline proxy.go]: @sigua-cs should I remove this comment as I add the decrement in line 255?

**Comment by mga-chka** [comment]
Thanks for the PR.
FYI, since your PR is linked to the tricky issues we had with the retry mechanism, we will take a bit of time to review it (since we will need multiple reviewer and because of the holidays it will take more time).

File: proxy.go
Comment (Blokje5) [inline proxy.go:249]: It might also a be a good idea to link the PR.

**Review by Blokje5 — APPROVED** [review]
LGTM! Great find.

File: proxyretry_test.go
Comment (mga-chka) [inline proxyretry_test.go]: You should use the **penaltySize** variable instead of 5, it would make the test more readable

File: proxy.go
Comment (mga-chka) [reply proxy.go]: yes you should since your PR fix this legacy quick & dirty (and broken) fix.

**Review by mga-chka — APPROVED** [review]
thanks for the PR, sorry for the long delay (we take a lot of days off in summer in France).
I left a few comments but they're very minor. Please tell me we want to fix them so that we don't wait for an extra 2 weeks before we merge the PR.

Once the PR is merged, we will release a new version ASAP so that you can benefit from your fix.

File: proxy.go
Comment (nir3c) [reply proxy.go]: done

File: proxy.go
Comment (nir3c) [reply proxy.go:249]: done

File: proxyretry_test.go
Comment (nir3c) [reply proxyretry_test.go]: done

**Comment by nir3c** [comment]
>

**Comment by nir3c** [comment]
> thanks for the PR, sorry for the long delay (we take a lot of days off in summer in France). I left a few comments but they're very minor. Please tell me we want to fix them so that we don't wait for an extra 2 weeks before we merge the PR.
> 
> Once the PR is merged, we will release a new version ASAP so that you can benefit from your fix.

thanks for reviewing the code, I update the code based on the comments you gave me, feel free review it again

**Review by mga-chka — APPROVED** [review]

**Comment by mga-chka** [comment]
> > thanks for the PR, sorry for the long delay (we take a lot of days off in summer in France). I left a few comments but they're very minor. Please tell me we want to fix them so that we don't wait for an extra 2 weeks before we merge the PR.
> > Once the PR is merged, we will release a new version ASAP so that you can benefit from your fix.
> 
> thanks for reviewing the code, I update the code based on the comments you gave me, feel free review it again

We should release a new version containing your fix on Monday.

**Pull request merged**
